File: exercism/python/saddle-points/saddle_points.py

def saddle_points(M):
    if len(set(map(len, M))) > 1:
        raise ValueError

    MT = list(zip(*M))
    points = set()
    for i, row in enumerate(M):
        for j, x in enumerate(row):
            if x == max(row) and x == min(MT[j]):
                points.add((i, j))
    return points

File: exercism/javascript/saddle-points/saddle-points.js
const BaseMatrix = require('../matrix/matrix')

const max = (xs) => Math.max.apply(null, xs)
const min = (xs) => Math.min.apply(null, xs)

function* genSaddlePoints ({rows, columns}) {
  for (let i = 0; i < rows.length; i++) {
    const row = rows[i]
    for (let j = 0; j < row.length; j++) {
      const x = row[j]
      if (x === max(row) && x === min(columns[j])) {
        yield [i, j]
      }
    }
  }
}

class Matrix extends BaseMatrix {
  constructor (args) {
    super(args)
    this.saddlePoints = Array.from(genSaddlePoints(this))
  }
}

module.exports = Matrix
